src/libs/elektra/key.c

```diff
@@ -626,3 +762,87 @@ ssize_t keyGetRef (const Key * key)
 
 	return key->ksReference;
 }
+
+
+/**
+ * @internal
+ * @see keyLock + keyIsLocked
+ */
+static int elektraKeyLock (Key * key, option_t what, int test)
+{
+	int ret = 0;
+
+	if (!key) return -1;
+
+	if (test_bit (what, KEY_LOCK_NAME))
+	{
+		if (!test_bit (key->flags, KEY_FLAG_RO_NAME))
+		{
+			if (!test) set_bit (key->flags, KEY_FLAG_RO_NAME);
+			set_bit (ret, KEY_LOCK_NAME);
+		}
+	}
+
+	if (test_bit (what, KEY_LOCK_VALUE))
+	{
+		if (!test_bit (key->flags, KEY_FLAG_RO_VALUE))
+		{
+			if (!test) set_bit (key->flags, KEY_FLAG_RO_VALUE);
+			set_bit (ret, KEY_LOCK_VALUE);
+		}
+	}
+
+	if (test_bit (what, KEY_LOCK_META))
+	{
+		if (!test_bit (key->flags, KEY_FLAG_RO_META))
+		{
+			if (!test) set_bit (key->flags, KEY_FLAG_RO_META);
+			set_bit (ret, KEY_LOCK_META);
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Permanently locks a part of the key
+ *
+ * This can be:
+ * - KEY_LOCK_NAME to lock the name
+ * - KEY_LOCK_VALUE to lock the value
+ * - KEY_LOCK_META to lock the metadata
+ *
+ * To unlock the key, duplicate it.
+ *
+ * It is also possible to lock when the key is created with
+ * keyNew().
+ *
+ * Some data structures need to lock the key (most likely
+ * its name), so that the ordering does not get confused.
+ *
+ * @param key which name should be locked
+ *
+ * @see keyNew(), keyDup(), ksAppendKey()
+ * @retval >0 the bits that were successfully locked
+ * @retval 0 if everything was locked before
+ * @retval -1 if it could not be locked (nullpointer)
+ * @ingroup key
+ */
+int keyLock (Key * key, option_t what)
+{
+	return elektraKeyLock (key, what, 0);
+}
+
+/**
+ * @brief Tests if a part of a key is locked
+ *
+ * @see keyLock() for details and return values
+ * @retval >0 the bits that were successfully locked
+ * @retval 0 if everything was locked before
+ * @retval -1 if it could not be locked (nullpointer)
```

Comment:
> if we ensure KEY_LOCK_* == KEY_FLAG_RO_*

Since it seems doing that might not be possible, a simple transform function that transforms `KEY_LOCK_*` into `KEY_FLAG_RO_*` would be needed. Something like:

```c
static int elektraLockFlagsToReadOnlyFlags (int lockFlags)
{
    keyflag_t readOnlyFlags;
    if (test_bit (lockFlags, KEY_LOCK_NAME)) set_bit (readOnlyFlags, KEY_FLAG_RO_NAME);
    if (test_bit (lockFlags, KEY_LOCK_VALUE)) set_bit (readOnlyFlags, KEY_FLAG_RO_VALUE);
    if (test_bit (lockFlags, KEY_LOCK_META)) set_bit (readOnlyFlags, KEY_FLAG_RO_META);
    return readOnlyFlags;
}
```

Since the bits are consecutive and in the same order in both cases this would also work, but it requires the magic number 16 (the difference in the positions of the bits):

```c
keyflag_t readOnlyFlags = (lockFlags & (KEY_LOCK_NAME | KEY_LOCK_VALUE | KEY_LOCK_META)) >> 16;
```

Reply:
> If I understand it correctly right now, the method returns that are NOT locked, which is kind of the opposite of what the name suggests

Yeah. Fixed that.